connaisseur/alert.py

```diff
@@ -0,0 +1,177 @@
+            "connaisseur_pod_id": os.getenv("POD_NAME"),
+            "cluster": load_config()["cluster_name"],
+            "timestamp": datetime.now(),
+            "request_id": admission_request.get("request", {}).get(
+                "uid", "No given UID"
+            ),
+            "images": (str(get_images(admission_request)) or "No given images"),
+        }
+        self.admission_request = admission_request
+        self.receiver_url = receiver_config["receiver_url"]
+        self.template = receiver_config["template"]
+        self.throw_if_alert_sending_fails = receiver_config.get(
+            "fail_if_alert_sending_fails", True
+        )
+        self.payload = self._construct_payload(receiver_config)
+        self.headers = get_headers(receiver_config)
+
+    def _construct_payload(self, receiver_config):
+        try:
+            with open(
+                "{}/{}.json".format(os.getenv("ALERT_CONFIG_DIR"), self.template),
+                "r",
+            ) as templatefile:
+                template = json.load(templatefile)
+        except Exception as err:
+            raise ConfigurationError(
+                "Template file for alerting payload is either missing or invalid JSON: {}".format(
+                    str(err)
+                )
+            )
+        payload = sanitize(template, self.context)
+        if receiver_config.get("payload_fields") is not None:
+            payload.update(receiver_config.get("payload_fields"))
+        return payload
+
+    def send_alert(self):
+        try:
+            response = requests.post(
+                self.receiver_url, data=self.payload, headers=self.headers
+            )
+            response.raise_for_status()
+            logging.info("sent alert to %s", self.template.split("_")[0])
+        except Exception as err:
+            if self.throw_if_alert_sending_fails:
+                raise AlertSendingError(str(err))
+            logging.error(err)
+        return response
+
+
+def load_config():
+    with open(
+        "{}/alertconfig.json".format(os.getenv("ALERT_CONFIG_DIR")), "r"
+    ) as configfile:
+        return json.loads(configfile.read())
+
+
+def get_images(admission_request):
+    relevant_spec = get_container_specs(
+        admission_request.get("request", {}).get("object", {})
+    )
+    return [container.get("image") for container in relevant_spec]
+
+
+def get_headers(receiver_config):
+    headers = {"Content-Type": "application/json"}
+    additional_headers = receiver_config.get("custom_headers")
+    if additional_headers is not None:
+        for header in additional_headers:
+            key, value = header.split(":", 1)
+            headers.update({key.strip(): value.strip()})
+    return headers
+
+
+def send_alerts(admission_request, *, admitted):
+    alert_config = load_config()
+    event_category = "admit_request" if admitted else "reject_request"
+    if alert_config.get(event_category) is not None:
```

Comment:
when a user submits a faulty `alert_config` he wouldn't get any feedback except for the alert messages never arriving, as there is no logging/exception risen that the `alert_config` has an invalid format.